scripts/compile-igs.js

```diff
@@ -0,0 +1,158 @@
+ *
+ */
+class IGsCompileHelper {
+    constructor(options = {}) {
+        this.options = options;
+    }
+
+    async listIgDirs(parentDir) {
+        return (await readDirPmd(parentDir, { withFileTypes: true }))
+            .filter(dirent => dirent.isDirectory())
+            .map(dirent => path.join(parentDir.toString(), dirent.name));
+    }
+
+    async compileIg(igDir) {
+        const indexJson = path.join(igDir.toString(), '.index.json');
+        if (!existsSync(indexJson)) {
+            return;
+        }
+        const index = await loadJson(indexJson);
+        index.files.forEach(file => {
+            if (file.resourceType === 'SearchParameter') {
+                const filePath = path.join(igDir.toString(), file.filename);
+                console.log(`Compiling ${filePath}`);
+                loadJson(filePath).then(jsonData => {
+                    compile(jsonData);
+                });
+            }
+        });
+    }
+
+    /**
+     * Main public function
+     * @param igsDir
+     */
+    async compileIGs(igsDir) {
+        await this.checkDependencies(igsDir);
+        for (const igPath of await this.listIgDirs(igsDir)) {
+            await this.compileIg(igPath);
+        }
+    }
+
+    createIGKey(name, version) {
+        if (this.options.ignoreVersion) {
+            return name;
+        }
+        return `${name}@${version}`;
+    }
+
+    async extractIgInfo(igDir) {
+        const packagePath = path.join(igDir.toString(), 'package.json');
+        if (!existsSync(packagePath)) {
+            return;
+        }
+        console.log(`checking ${packagePath}`);
+        const packageJson = await loadJson(packagePath);
+        const dependencies = [];
+        const retVal = {
+            id: this.createIGKey(packageJson.name, packageJson.version),
+            url: packageJson.url,
+            name: packageJson.name,
+            version: packageJson.version,
+            dependencies,
+        };
+        const packageDeps = packageJson.dependencies;
+        if (packageDeps) {
+            for (const [name, version] of Object.entries(packageDeps)) {
+                const igId = this.createIGKey(name, version);
+                dependencies.push(igId);
+            }
+        }
+        return retVal;
+    }
+
+    async checkDependencies(igsDir) {
+        const parentMap = {};
+        for (const igPath of await this.listIgDirs(igsDir)) {
+            console.log(`looking at ig path: ${igPath}`);
+            const igInfo = await this.extractIgInfo(igPath);
+            if (igInfo) {
+                parentMap[igInfo.id] = igInfo.dependencies;
+            }
+        }
+        for (const igId of Object.keys(parentMap)) {
+            this.depthFirst([igId], parentMap);
+        }
+    }
+
+    depthFirst(parents, pMap) {
+        const igId = parents[parents.length - 1];
+        const dependencies = pMap[igId];
+        if (!dependencies) {
+            throw new Error(`Missing dependency ${igId}`);
```

Comment:
We need to make an exception for `hl7.fhir.r4.core`. There's no need to compile the base FHIR definitions since they are already already bundled in FWoA.